snapcraft/providers/_get_provider.py

```diff
@@ -28,40 +32,57 @@
 def get_provider(provider: Optional[str] = None) -> Provider:
     """Get the configured or appropriate provider for the host OS.
 
-    If platform is not Linux, use Multipass.
-
-    If platform is Linux:
+    To determine the appropriate provider,
     (1) use provider specified in the function argument,
-    (2) use provider specified with snap configuration if running
-        as snap,
-    (3) get the provider from the environment if valid,
-    (4) default to platform default (LXD on Linux).
+    (2) get the provider from the environment if valid,
+    (3) use provider specified with snap configuration,
+    (4) default to platform default (LXD on Linux, otherwise Multipass).
 
     :return: Provider instance.
     """
+    chosen_provider = ""
     env_provider = os.getenv("SNAPCRAFT_BUILD_ENVIRONMENT")
-    env_provider_is_valid = env_provider in ("lxd", "multipass")
 
-    if provider is None and env_provider_is_valid:
-        provider = env_provider
-    elif provider is None:
-        provider = get_platform_default_provider()
-
-    if provider == "lxd":
+    # load snap config file
+    snap_config = get_snap_config()
+    if snap_config:
+        snap_provider = snap_config.provider
+    else:
+        snap_provider = None
+
+    # (1) use provider specified in the function argument
+    if provider:
+        emit.debug(f"Using provider {provider!r} passed as an argument.")
+        chosen_provider = provider
+
+    # (2) get the provider from the environment if valid
+    elif env_provider:
+        emit.debug(
+            f"Using provider {env_provider!r} from environmental "
+            "variable 'SNAPCRAFT_BUILD_ENVIRONMENT'."
+        )
+        chosen_provider = env_provider
+
+    # (3) use provider specified with snap configuration
+    elif snap_provider:
+        emit.debug(f"Using provider {snap_provider!r} from snap config.")
+        chosen_provider = snap_provider
+
+    # (4) default to platform default (LXD on Linux, otherwise Multipass)
+    elif sys.platform == "linux":
+        emit.debug("Using default provider 'lxd' on linux system.")
+        chosen_provider = "lxd"
+    else:
+        emit.debug("Using default provider 'multipass' on non-linux system.")
+        chosen_provider = "multipass"
+
+    # ignore case
+    chosen_provider = chosen_provider.lower()
+
+    if chosen_provider not in ("lxd", "multipass"):
+        raise RuntimeError(f"unsupported provider specified: {chosen_provider!r}")
```

Comment:
Too generic. As the value is wrong, what about using `ValueError`?

Reply:
Used `ValueError`.  I also made the if statements more pythonic.